tests/unit_tests/plugin_manager_test.py

```diff
@@ -0,0 +1,169 @@
+    re.compile("direct_pattern"): expand_func_names(FakePlugin.direct_function),
+    re.compile("another_async_pattern"): expand_func_names(
+        FakePlugin.my_async_function
+    ),
+    re.compile("async_pattern"): expand_func_names(FakePlugin.my_async_function),
+    re.compile("hi_custom"): expand_func_names(FakePlugin.hi_custom),
+    # Click commands construct a regex pattern from the listen_to pattern
+    re.compile("^click_command (.*)?"): expand_func_names(FakePlugin.click_commmand),
+}
+
+hook_listeners = {
+    re.compile("webhook_id"): expand_func_names(FakePlugin.webhook_listener)
+}
+
+
+class TestPlugin:
+    def test_init_plugin(self):
+        p = FakePlugin()
+        m = PluginManager([p])
+        with mock.patch.object(p, "initialize") as mocked:
+            m.initialize(Driver(), Settings())
+
+            mocked.assert_called_once()
+
+    def test_initialize(self):
+        m = PluginManager([FakePlugin()])
+        m.initialize(Driver(), Settings())
+
+        # Test whether the function was registered properly
+        assert m.message_listeners[re.compile("pattern")] == [
+            FakePlugin.my_function,
+        ]
+
+        # This function should be registered twice, once for each listener
+        assert len(m.message_listeners[re.compile("async_pattern")]) == 1
+        assert (
+            m.message_listeners[re.compile("async_pattern")][0].function
+            == FakePlugin.my_async_function.function
+        )
+
+        assert len(m.message_listeners[re.compile("another_async_pattern")]) == 1
+        assert (
+            m.message_listeners[re.compile("another_async_pattern")][0].function
+            == FakePlugin.my_async_function.function
+        )
+
+        assert len(m.webhook_listeners) == 1
+        assert m.webhook_listeners[re.compile("webhook_id")] == [
+            FakePlugin.webhook_listener
+        ]
+
+
+class TestPluginManager:
+    def setup_method(self):
+        self.p = FakePlugin()
+        self.plugin_manager = PluginManager([self.p])
+
+    def test_init(self):
+        self.plugin_manager.initialize(Driver(), Settings())
+
+        # Test that listeners of individual plugins are now registered on the plugin_manager
+        assert len(msg_listeners) == len(self.plugin_manager.message_listeners)
+        for pattern, listeners in self.plugin_manager.message_listeners.items():
+            for listener in listeners:
+                assert pattern in msg_listeners
+                assert listener in msg_listeners[pattern]
+
+        assert len(hook_listeners) == len(self.plugin_manager.webhook_listeners)
+        for pattern, listeners in self.plugin_manager.webhook_listeners.items():
+            for listener in listeners:
+                assert pattern in hook_listeners
+                assert listener in hook_listeners[pattern]
+
+    def test_get_help(self):
+        # Prior to initialization there is no help
+        assert self.plugin_manager.get_help() == []
```

Comment:
This seems to rely on `self.plugin_manager` not having been initialized, but it's initialized first thing in `test_init`?

Reply:
Doesn't pytest run the two tests independently? i.e.
```
t = TestPluginManager()
t.setup_method()
t.test_get_help()

t = TestPluginManager()
t.setup_method()
t.test_get_init()
```